Use a RichText component for the File block's filename editable content

talldan opened the pull request:
## Description
Closes #8023 

File block was using a `contenteditable` anchor element for its editor link text. This replaces the anchor element's content with a `RichText`

## How has this been tested?
Manual testing

## Types of changes
<!-- Bug fix (non-breaking change which fixes an issue) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

talldan committed:
Make editable contents of FileBlockEditableLink a RichText

talldan committed:
Prevent the anchor element's default click behaviour

talldan commented:
@noisysocks - I wonder if it's worth getting rid of the anchor element entirely and not worrying about the copy/paste behaviour. It's a bit weird having to put a div inside of an anchor tag. Also, preventing the anchor link from opening is weird.

I've also noticed something weird going on with file blocks - I sometimes get a message 'This block appears to have been modified externally.' when it hasn't. Eeek.

noisysocks commented on core-blocks/file/editable-link.js:
```diff
@@ -39,40 +29,31 @@ export default class FileBlockEditableLink extends Component {
-					{ text }
+					<RichText
+						wrapperClassName={ `${ className }__textlink-richtext` }
+						tagName="div" // must be block-level or else cursor disappears
```
Why does the cursor disappear?

noisysocks commented:
Yeah, I think you're right. It feels really unintuitive having the cursor be a hand instead of a vertical bar.

talldan committed:
Remove File Block's editable link component in favour of the RichText component

talldan committed:
Convert fileName attribute to string for use as element attribute

Since fileName is a value set by a RichText element, we need to convert it back
to a string for use as an element attribute.

This has the side-effect that the download button's download attribute is
always set, even when the string is empty, but this results in the correct
behaviour of the download button still causing a download of the file. In this
case the browser reverts back to the default filename.

talldan committed:
Ensure optional rendering of filename link continues to works with RichText value

talldan committed:
Use RichText format="string" instead of renderToString

noisysocks commented on core-blocks/file/index.js:
```diff
@@ -217,7 +217,7 @@ export const settings = {
 						href={ href }
 						className="wp-block-file__button"
-						download={ fileName }>
+						download={ fileName || '' }>
```
Why is this needed?

talldan replied:
I had an issue before whereby if the fileName is undefined, the download attribute is not added to the resulting anchor, which changes the behaviour of the button from downloading a file to opening it in a new tab. It's probably not needed now. Not sure if there could ever be an edge-case where a fileName could be undefined now that the string format is being used for the rich text field.

noisysocks replied:
Fair enough. I wonder if it should be `fileName || true` so that if there is no file name, the attribute is added but not set to anything? e.g. `<a href="..." download>`

noisysocks reviewed (approved):
Love it! 😍

talldan committed:
Add clarifying comment
